### ISSUE TITLE
DOC: kmeans and kmeans2

### ISSUE THREAD

**Issue by MartinThoma** [scipy/scipy#7250]
It is not clear to me from the documentation when I should use [`kmeans`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.vq.kmeans.html#scipy.cluster.vq.kmeans) and when [`kmeans2`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.vq.kmeans2.html#scipy.cluster.vq.kmeans2). The [`cluster.vq` module docs](https://docs.scipy.org/doc/scipy/reference/cluster.vq.html#module-scipy.cluster.vq) do not clarify that either. There is also a [StackOverflow question](http://stackoverflow.com/q/20661142/562769) about this.

I think at least the names of the parameters and their description could be made consistent:

| kmeans        | kmeans2       |
| ------------- | ------------- |
| obs           | data          |
| k_or_guess    | k             |
| iter          | iter          |
| thres         | thres         |
| -             | missing       |
| check_finite  | check_finite  |
| codebook      | centroid      |
| distortion    | -             |
| -             | label         |

There is only the note

> a different implementation of k-means clustering with more methods for generating initial centroids but without using a distortion change threshold as a stopping criterion.

in kmeans. Just looking at the documentation, it seems to me that it might be possible to join those two functions to one. If missing is used, then kmeans2 is used.

**Comment by perimosocordiae** [comment]
I'm +1 for merging their functionality into `kmeans` and deprecating `kmeans2`. There's also a lot of room for optimization.

**Comment by rgommers** [comment]
> There's also a lot of room for optimization.

Indeed. Here's a nice benchmark: http://hdbscan.readthedocs.io/en/latest/performance_and_scalability.html#comparison-of-high-performance-implementations. The scikit-learn implementation is a lot faster.

**Comment by mdhaber** [comment]
@rgommers You may have noticed the recent issues #8683 and #8727. Besides fixing these bugs, do you think `scipy.cluster.kmeans` is something that should be improved further (like by addressing this issue and perhaps finishing #4119), or should we continue to offer it for convenience and refer users to scikit-learn for performance? @perimosocordiae I see you've been active on all the `kmeans` issues. What do you think?

**Comment by rgommers** [comment]
I'd say merging the recent PRs and closing this issue by merging the implementations make sense. gh-4119 could be merged if someone picks it up, but that's already in the "new features" domain. I'd say simple PRs that make things better should continue to be accepted, but I'd say for nonstandard functionality and optimal performance we should just refer users to scikit-learn.

**Comment by Liamolucko** [comment]
> Indeed. Here's a nice benchmark: http://hdbscan.readthedocs.io/en/latest/performance_and_scalability.html#comparison-of-high-performance-implementations. The scikit-learn implementation is a lot faster.

This benchmark doesn't seem very fair to me - since it uses the default settings for sklearn and scipy, which default to 1 and 20 iterations respectively, scipy has to do 20 times as much work as sklearn does in that benchmark.

I tweaked the benchmark to use the same number of iterations for both, and also include `kmeans2` - it turns out that scipy's `kmeans` is legitimately slower than sklearn's, but `kmeans2` is actually faster than sklearn!

<img width="1311" height="942" alt="Image" src="https://github.com/user-attachments/assets/f8a7089b-ad89-48ea-8e42-ab65ff963390" />

<details>

<summary>Modified benchmark</summary>

```python
import sklearn.cluster
import scipy.cluster
import sklearn.datasets
import numpy as np
import pandas as pd
import time
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_context('poster')
sns.set_palette('Paired', 10)
sns.set_color_codes()

def benchmark_algorithm(dataset_sizes, cluster_function, function_args, function_kwds,
                        dataset_dimension=10, dataset_n_clusters=10, max_time=45, sample_size=2):

    # Initialize the result with NaNs so that any unfilled entries
    # will be considered NULL when we convert to a pandas dataframe at the end
    result = np.nan * np.ones((len(dataset_sizes), sample_size))
    for index, size in enumerate(dataset_sizes):
        for s in range(sample_size):
            # Use sklearns make_blobs to generate a random dataset with specified size
            # dimension and number of clusters
            data, labels = sklearn.datasets.make_blobs(n_samples=size,
                                                       n_features=dataset_dimension,
                                                       centers=dataset_n_clusters)

            # Start the clustering with a timer
            start_time = time.time()
            cluster_function(data, *function_args, **function_kwds)
            time_taken = time.time() - start_time

            # If we are taking more than max_time then abort -- we don't
            # want to spend excessive time on slow algorithms
            if time_taken > max_time:
                result[index, s] = time_taken
                return pd.DataFrame(np.vstack([dataset_sizes.repeat(sample_size),
                                               result.flatten()]).T, columns=['x','y'])
            else:
                result[index, s] = time_taken

    # Return the result as a dataframe for easier handling with seaborn afterwards
    return pd.DataFrame(np.vstack([dataset_sizes.repeat(sample_size),
                                   result.flatten()]).T, columns=['x','y'])

dataset_sizes = np.hstack([np.arange(1, 6) * 500, np.arange(3,7) * 1000, np.arange(4,17) * 2000])

large_dataset_sizes = np.arange(1,16) * 4000

k_means_default = sklearn.cluster.KMeans(10)
large_k_means_default_data = benchmark_algorithm(large_dataset_sizes,
                                         k_means_default.fit, (), {},
                                         max_time=90, sample_size=1)

k_means = sklearn.cluster.KMeans(10, n_init=10)
large_k_means_data = benchmark_algorithm(large_dataset_sizes,
                                         k_means.fit, (), {},
                                         max_time=90, sample_size=1)

large_scipy_k_means_default_data = benchmark_algorithm(large_dataset_sizes,
                                               scipy.cluster.vq.kmeans, (10,), {},
                                               max_time=90, sample_size=1)

large_scipy_k_means_data = benchmark_algorithm(large_dataset_sizes,
                                               scipy.cluster.vq.kmeans, (10,), {"iter": 10},
                                               max_time=90, sample_size=1)

large_scipy_k_means_2_default_data = benchmark_algorithm(large_dataset_sizes,
                                               scipy.cluster.vq.kmeans2, (10,), {},
                                               max_time=90, sample_size=1)

large_scipy_k_means_2_data = benchmark_algorithm(large_dataset_sizes,
                                               scipy.cluster.vq.kmeans2, (10,), {"iter": 10, "minit": "++"},
                                               max_time=90, sample_size=1)

sns.regplot(x='x', y='y', data=large_k_means_default_data, order=2,
            label='Sklearn K-Means (Default, 1 Iter)', x_estimator=np.mean)
sns.regplot(x='x', y='y', data=large_k_means_data, order=2,
            label='Sklearn K-Means (10 Iters)', x_estimator=np.mean)
sns.regplot(x='x', y='y', data=large_scipy_k_means_default_data, order=2,
            label='Scipy K-Means (Default, 20 Iters)', x_estimator=np.mean)
sns.regplot(x='x', y='y', data=large_scipy_k_means_data, order=2,
            label='Scipy K-Means (10 Iters)', x_estimator=np.mean)
sns.regplot(x='x', y='y', data=large_scipy_k_means_2_default_data, order=2,
            label='Scipy K-Means 2 (Default, 10 Iters, Random)', x_estimator=np.mean)
sns.regplot(x='x', y='y', data=large_scipy_k_means_2_data, order=2,
            label='Scipy K-Means 2 (10 Iters, K-Means++)', x_estimator=np.mean)

plt.gca().axis([0, 64000, 0, 1])
plt.gca().set_xlabel('Number of data points')
plt.gca().set_ylabel('Time taken to cluster (s)')
plt.title('Performance Comparison of Fastest Clustering Implementations')
plt.legend()

plt.show()
```

</details>
